corehq/apps/translations/static/translations/js/translations.js

```diff
@@ -17,7 +17,11 @@ hqDefine("translations/js/translations", function () {
                 var Translation = function (key, value) {
                     var that = this;
```

Comment:
Totally unrelated: What's the point of `that = this`?

Reply:
It's common to reassign `this` to a variable in javascript because it's easy to lose track of the meaning of `this` when you're dealing with callbacks. My preference is to always reassign `this` and then exclusively reference the variable, although in single-line functions directly referencing `this` is okay. Officially we're not that strict about it, we could add a rule like [this](https://eslint.org/docs/rules/consistent-this) to be stricter.

Looking at this file more closely, I don't like that it uses a mix of `this` and `that`. Updating this in 969550ad0db Also renaming `that` to `self` which is more common in our code.

Reply:
That makes a lot of sense, thanks for the explanation!

Reply:
@gbova JavaScript has weird semantics around the value of `this` in a function context. For example:

```js
function Thing() {
    this.value = 1;
    var that = this;

    this.getCallback = function () {
        return function (value) {
            // BAD: `this` may not reference `thing`,
            // depending on how this function is called
            //return this.value + value

            // `that` will always reference the instance of `thing`
            // on which `getCallback` was called
            return that.value + value
        }
    }
}

var thing = new Thing();
var callback = thing.getCallback();

//  here `this` would be either `undefined` or `window`
// (depending on strict mode) inside `callback`
callback(2);

// here `this` would reference `thing`
// (although this is not a typical way to call a function)
// the first argument passed to `apply` will be the value of `this` in `callback`
callback.apply(thing, 2);
```

Here's a better, and longer, explanation of [how `this` works](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this).

Reply:
Hehe, took too long to write my answer and Jenny beat me to it.